includes/admin/class-wc-rest-payments-orders-controller.php

```diff
@@ -217,6 +226,30 @@ public function capture_terminal_payment( WP_REST_Request $request ) {
 			}
 			// Store receipt generation URL for mobile applications in order meta-data.
 			$order->add_meta_data( 'receipt_url', get_rest_url( null, sprintf( '%s/payments/readers/receipts/%s', $this->namespace, $intent->get_id() ) ) );
+
+			// Add payment method for future subscription payments.
+			$generated_card = $intent->get_charge()->get_payment_method_details()[ Payment_Method::CARD_PRESENT ]['generated_card'] ?? null;
+			// If we don't get a generated card, e.g. because a digital wallet was used, we can still return that the initial payment was successful.
+			// The subscription will not be activated and customers will need to provide a new payment method for renewals.
+			if ( $generated_card ) {
+				$has_subscriptions = function_exists( 'wcs_order_contains_subscription' ) &&
+										function_exists( 'wcs_get_subscriptions_for_order' ) &&
+										function_exists( 'wcs_is_manual_renewal_required' ) &&
+										wcs_order_contains_subscription( $order_id );
+				if ( $has_subscriptions ) {
+					$token = $this->token_service->add_payment_method_to_user( $generated_card, $order->get_user() );
+					$this->gateway->add_token_to_order( $order, $token );
+					foreach ( wcs_get_subscriptions_for_order( $order ) as $subscription ) {
+						$subscription->set_payment_method( WC_Payment_Gateway_WCPay::GATEWAY_ID );
+						// Where the setting doesn't force manual renewals, we should turn them off, because we have an auto-renewal token now.
+						if ( ! wcs_is_manual_renewal_required() ) {
+							$subscription->set_requires_manual_renewal( false );
+						}
+						$subscription->save();
+					}
+				}
+			}
+
```

Comment:
Thanks for your review @zmaglica. For context, I work as a mobile engineer and I'm finding my feet with PHP when it makes sense... so apologies if some of my questions are a bit basic!

If we move the code to the order service, we'll add a dependency between the order and token services. Do we want that? The service would also need the gateway in order to add the token. Perhaps we could pass the token creation/addition in a closure to avoid this coupling...

However, this code is quite specific to part of the API used by the mobile app – when the same thing happens on the web, it's handled through the cart and checkout by the subscriptions plugin. It seems unlikely we'd want to use this outside of the REST controller in future. Given that this code "fills a gap" left in the API by the subscriptions plugin, perhaps the REST controller is the right place for it... WDYT?

I'll experiment a little with the changes anyway.

> One thing worth clarifying is the urgency of this PR for Mobile teams; @joshheald, can you share insights, please?
> – https://github.com/Automattic/woocommerce-payments/pull/8390#issuecomment-2079317192

My team will soon be shifting focus to a different project with some urgency, and may hand responsibility for this project over to another team when that happens. In order to do that smoothly, it would be really helpful to merge this sooner rather than later. If you don't have major concerns, perhaps we could continue and improve it in future should the need arise?